data/transactions/logic/assembler.go

```diff
@@ -1545,9 +1548,16 @@ func (ops *OpStream) trackStack(args StackTypes, returns StackTypes, instruction
 	}
 }
 
-// splitTokens breaks tokens into two slices at the first semicolon.
-func splitTokens(tokens []string) (current, rest []string) {
-	for i, token := range tokens {
+// splitTokens breaks tokens into two slices at the first semicolon and expands macros along the way.
+func splitTokens(ops *OpStream, tokens []string) (current, rest []string) {
+	for i := 0; i < len(tokens); i++ {
+		token := tokens[i]
+		replacement, ok := ops.macros[token]
+		if ok {
+			tokens = append(tokens[0:i], append(replacement, tokens[i+1:]...)...)
```

Comment:
Further down, I ask "why copy"? The reason you had to do it, I guess, is here. This append could modify the backing array of `replacement`. The only reason you don't have a bug is because you made a slice copy of exactly the right size, so this append always reallocates rather than extends.  Top force that without the copy, use the fill slice notation. Set the expansion to `tokens[2:len(tokens):len(tokens-2)]` rather than make the copy.

Reply:
And, by the way, the first `append` can mess with `tokens` after `i`.

Reply:
Ah, the capacity thing's much cleaner; good point! I'm not understanding what you mean about the first append; don't we want to change tokens after i? The inner append is going to have a new array if tokens[i+1:] has non-zero length, so we don't care if the outer append stomps over tokens.

Reply:
OK, I missed the purpose of the `i--`. That seems food.  Could you `// backup to process the expansion` or similar?

Reply:
This does mean that expansion to nothing isn't allowed. I know you check for that in `#define`, so I think it's probably fine.

But then again, it could be useful. Imagine, for example
```
#define DEBUG#1 byte #!; log
```
And the author changes that to:
```
#define DEBUG#1
```
before assembling for deployment

Reply:
I think we're good now. We know that `replacement` has a tight capacity, so the inner append creates a new backing store, copying the end of the `tokens` slice.  Then it appends into `tokens`, which might stomp on the rest, but that's ok - in fact it's the intent.